### DISCUSSION TITLE
Execution order of dependencies

### DISCUSSION THREAD

**Discussion by Soerboe** [fastapi/fastapi#6294] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uvicorn as uvicorn
from fastapi import FastAPI, Depends

app = FastAPI()


def dep1():
    print('1')


def dep2():
    print('2')


@app.get('/test', dependencies=[Depends(dep1), Depends(dep2)])
def test():
    return 'test'


if __name__ == '__main__':
    uvicorn.run("main:app", host="127.0.0.1", port=8008, reload=True, debug=True)
```

### Description

By running the above code the output is 

```
1
2
```

Is the execution order of dependencies always the declared order? I haven't found anything in the docs regarding the order of the dependencies.

### Operating System

Linux

### FastAPI Version

0.73.0

### Python Version

3.10.6

**Comment by iudeen** [comment]
They must be ordered since you're passing them on a list

**Comment by Soerboe** [comment]
Well, the order from my side is fixed but that does not guarantee that the execution order of the dependencies are the same. That is totally up to the implementation in FastAPI. My question is therefore whether FastAPI guarantees that the declared order is the same as execution order?

**Comment by imacat** [comment]
I suppose the order is from left to right.  That means, dep1, and then dep2.  However, the order is not guaranteed.
And the order is dealt with by the Python compiler, not FastAPI.

**Comment by Soerboe** [comment]
@imacat  The order of the elements in a list is guaranteed by python. A list is an ordered collection. Please, check your statements before posting.

However, FastAPI could choose to reorder the `Depends()` in the list for some reason. I want to know whether FastAPI would do this or not.

**Comment by imacat** [comment]
@Soerboe OK, you are right.  I would shut-up and speak no more.

**Comment by JarroVGIT** [comment, marked as answer]
I suppose any current order effects ar an implementation and not a guarantee. The only guarantee that exists is that a ‘Depends()’ that itself refers to a ‘Depends()’ will have the latter execute first (as the result is needed in the original dependency).

**Comment by zhanymkanov** [comment]
If you want to guarantee the order of dependencies you can chain them, i.e. second dependency calls the first one.
Find an example [here](https://github.com/zhanymkanov/fastapi-best-practices#4-chain-dependencies).

From my experience, dependencies were called in the order they have been declared whether it's a parameter in a route or a list of `dependencies=[]`

I could assume it as a non-official rule since python is an interpreted language and MRO rules in python are quite similar.

**Comment by Soerboe** [comment]
@JarroVGIT I see.

@zhanymkanov Nice best practices. Found a good solution to my problem in the chained dependencies section.

**Comment by Wyko** [comment]
@tiangolo 
Could you confirm if this is a permanent, dependable function of FastAPI? Specifically, that Dependencies will always be called in the order they are listed on the endpoint, except if they are chained somehow?
